src/chkstat.cpp

```diff
@@ -21,1365 +21,1246 @@
     {
-        if (!strcmp(level[i], "easy") || !strcmp(level[i], "secure") || !strcmp(level[i], "paranoid"))
+        if (file.find_first_of('.') != file.npos)
+            // we're only interested in base profiles
             continue;
 
-        std::string profile = etc_root + "/permissions." + level[i];
+        const auto path = dir + "/" + file;
 
-        if (access(profile.c_str(), R_OK) == 0)
+        // NOTE: we already know that the path exist(ed), because it was
+        // returned from the readdir() above. The existence check is racy
+        // anyways so no need to recheck.
+        m_profile_paths.push_back(path);
+
+        /*
+         * this is a bit of strange logic here, because we need to add the per
+         * package profile files in the order as the profiles appear in
+         * m_profiles.
+         */
+        for (const auto &profile: m_profiles)
         {
-            ensure_array((void**)&permfiles, &npermfiles);
-            permfiles[npermfiles++] = strdup(profile.c_str());
+            const auto profile_basename = file + "." + profile;
+
+            if (files.find(profile_basename) != files.end())
+            {
+                m_profile_paths.push_back(path + "." + profile);
+            }
         }
     }
 }
 
-
-void
-usage(int x)
+ProfileEntry&
+Chkstat::addProfileEntry(const std::string &file, const std::string &owner, const std::string &group, mode_t mode)
 {
-    printf("Usage:\n"
-           "a) chkstat [OPTIONS] <permission-files>...\n"
-           "b) chkstat --system [OPTIONS] <files>...\n"
-           "\n"
-           "Options:\n"
-           "  --set                 apply changes\n"
-           "  --warn                only tell which changes are needed\n"
-           "  --noheader            don't print intro message\n"
-           "  --fscaps              force use of fscaps\n"
-           "  --no-fscaps           disable use of fscaps\n"
-           "  --system              system mode, act according to /etc/sysconfig/security\n"
-           "  --level LEVEL         force use LEVEL (only with --system)\n"
-           "  --examine FILE        apply to specified file only\n"
-           "  --files FILELIST      read list of files to apply from FILELIST\n"
-           "  --root DIR            check files relative to DIR\n"
-           "  --config-root DIR     lookup config files relative to DIR\n"
-    );
-    exit(x);
+    // use the full path including a potential alternative root directory as
+    // map key
+    std::string path = m_root_path.getValue();
+    if (!path.empty())
+    {
+        path += '/';
+    }
+    path += file;
+
+    // this overwrites an possibly already existing entry
+    // this is intended behaviour, hence the order in which profiles are
+    // applied is important
+    auto &entry = m_profile_entries[path];
+
+    entry.file = path;
+    entry.owner = owner;
+    entry.group = group;
+    entry.mode = mode;
+    entry.caps.destroy();
```

Comment:
After checking again, that's actually the case. I guess the `destroy` is only necessary because we might be overriding an existing entry from a file that was loaded earlier.

So this is fine as-is.